landside/catkin_ws/src/gui/src/gui/system_widget.py

```diff
@@ -0,0 +1,144 @@
+        self.state_button.clicked.connect(lambda: self.execute_button_clicked('state'))
+
+        self.tasks_button.setText(f'Start {self.LAUNCHFILES["tasks"]}')
+        self.tasks_button.clicked.connect(lambda: self.execute_button_clicked('tasks'))
+
+        self.table_widget.setSelectionBehavior(QAbstractItemView.SelectRows)
+
+        self.launch_dialog = LaunchDialog()
+        self.launch_dialog.setObjectName('LaunchDialog')
+        self.launch_dialog.reset()
+        self.launch_dialog.node_launched.connect(self.append_to_table)
+
+        self.launch_dialog_button.clicked.connect(self.launch_node_dialog)
+
+        self.remote_launch_timer = QTimer(self)
+        self.remote_launch_timer.timeout.connect(self.check_remote_launch)
+        self.remote_launch_timer.start(100)
+
+        rospy.loginfo('System Widget successfully initialized')
+
+    def keyPressEvent(self, event):
+        if event.key() == QtCore.Qt.Key_Delete:
+            items = self.table_widget.selectedItems()
+            for key in self.execute_rows:
+                if items and self.execute_rows[key] == items[0].row():
+                    self.execute_button_clicked(key)
+                    items = None
+            if items and items[0].row() != 0:
+                self.delete_launch(items[0].row())
+        super(SystemWidget, self).keyPressEvent(event)
+
+    def check_remote_launch(self):
+        enabled = '/start_node' in rosservice.get_service_list()
+        self.launch_dialog_button.setEnabled(enabled)
+        self.table_widget.setEnabled(enabled)
+        if enabled and self.current_launch is None:
+            for key in self.execute_buttons:
+                self.execute_buttons[key].setEnabled(True)
+        elif enabled and self.current_launch is not None:
+            for key in self.execute_buttons:
+                self.execute_buttons[key].setEnabled(key == self.current_launch)
+        else:
+            for key in self.execute_buttons:
+                self.execute_buttons[key].setEnabled(False)
+
+    def check_system_sub(self):
+        self.system_usage_box.setEnabled(rospy.Time.now() - self.last_system_time < rospy.Duration(2))
+
+    def simulation_changed(self, val):
+        self.simulation = self.simulation_check_box.isChecked()
+
+    def update_cpu_ram(self, system_usage):
+        self.last_system_time = rospy.Time.now()
+        self.cpu_value.display(system_usage.cpu_percent)
+        self.ram_value.display(system_usage.ram.used)
+
+    def launch_node_dialog(self):
+        self.launch_dialog.reset()
+        self.launch_dialog.open()
+
+    def launch_file(self, launchfile):
+        args = ['sim:=true'] if self.simulation else ['sim:=false']
+        start_launch = rospy.ServiceProxy('start_node', StartLaunch)
+        resp = start_launch('execute', launchfile, args, True)
+        return self.append_to_table(resp.pid, 'execute', launchfile, args[0])
+
+    def delete_launch(self, row_value):
```

Comment:
Stopping any running launch file crashes the entire GUI.

Reply:
This still happens intermittently when running certain launch files. I haven't been able to find the root cause, but it doesn't seem to be an issue with the `remote_launch.py` script. The process shuts down correctly and delete_launch is also successful, but somewhere later down the line the gui segfaults. I really have no clue what it is. To reproduce:

1. Run `roslaunch system_utils remote_launch.launch`
2. Run `roslaunch gui gui.launch`
3. Run simulation and enable simulation mode in gui
4. Start `motion.launch` from gui
5. Publish a desired pose from gui
6. Enable & disable controls
7. Attempting to stop `motion.launch` crashes the gui

Something similar happens to tasks.launch. The only lead I have is that if there is an [error] message in the `roslaunch system_utils remote_launch.launch` tab during shutdown the gui is more likely to crash, but this isn't always the case.

Reply:
Yeah, I wasn't able to figure out what was going on last year, I recommend checking out the logs located under ~/.ros/logs or something like that

Reply:
Because the gui is functional I'm opting to resolve this conversation . I'll add a backlog item for fixing this bug. The CV team is working on widgets to show camera feeds + bounding boxes as well as a cv control panel.